lib/geocoder-geonames.js

```diff
@@ -23,6 +23,12 @@ function GeocoderGeonames (options) {
     this.endpoint = this.endpoint || this.endpoints.secure;
   }
 
+  if(this.options.premium && !this.options.token) {
```

Comment:
What you say is true, but in this case the premium option is useless.
It would be better to rename it "secure", I let you see this suggestion in my last commit.

Reply:
Ok, let's clarify the modes:

1. Premium means, you have a paid subscription at Geonames, this allows you to use the the Premium Hosts (http://ws.geonames.net or https://secure.geonames.net) or the premium IPs.
2. Secure means, you're premium and want to use the SSL endpoint.
3. Token means, you will use an optional token for authentication, this is a premium only feature.

So the conditions should be:

```
if (!this.options.username) {
  throw new Error('Please provide a valid username!');
}

if (this.options.token && !this.options.premium) {
  throw new Error('Token auth is for premium only!');
}

```

Or do I misunderstood something ?

Reply:
I think we don't really care if the user is premium or not.
What we need to know is if they want to use the secure url or not; if they want to use it, they need to give us a token.

Do you see what I mean?

Reply:
Ok, now I got it. But remember, all modes must be possible:

1. Free (api.geonames.org), no SSL, no Token
```
options = {
  username: USERNAME
});
````
2. Premium (ws.geonames.net), no SSL, no Token
```
options = {
  username: USERNAME,
  premium: true,
});
````
3. Premium Token (ws.geonames.net), no SSL
```
options = {
  username: USERNAME,
  premium: true,
  token: TOKEN,
});
```
3. Premium SSL (secure.geonames.net), no Token
```
options = {
  username: USERNAME,
  premium: true,
  secure: true,
});
```
4. Premium SSL+Token (secure.geonames.net)
```
options = {
  username: USERNAME,
  secure: true,
  premium: true,
  token: TOKEN,
});
```

With option.endpoint, you can overwrite the default endpoints, such as IP addresses. 
```
options = {
  username: USERNAME,
  premium: true,
  token: TOKEN,
  endpoint: 199.189.87.43
});
```

Reply:
For me the `ws` and `api` url are the same, I try to call the `ws` URL with a free account and it works.
And for the secure URL if you don't give any token, **it doesn't work**.

So the different possible cases for me are :

- Free (api.geonames.org), no SSL, no Token
```
options = {
  username: USERNAME,
});
```

- Premium SSL+Token (secure.geonames.net)
```
options = {
  username: USERNAME,
  secure: true,
  token: TOKEN,
});
```

And for the last point, with my change you can already send a custom endpoint if you want.

You can try on your side to validate this.

Reply:
> For me the ws and api url are the same, I try to call the ws URL with a free account and it works.

No, they're not. The ws endpoint is for premium only and has much faster response time.

> And for the secure URL if you don't give any token, it doesn't work.

Sure it does: Try https://secure.geonames.net/search?q=Berlin&username=YOUREUSERNAME

Reply:
It's strange because some call work with demo account :
- http://ws.geonames.org/get?geonameId=6690364&username=demo

It's the same for the secure one, if I call it without our token the call ask us to add the token.

So it depends on the calls, strange...
I will see to refactor my commit then, thanks 😃

Reply:
I can add you to the repo if you want.